cdap-api/src/main/java/co/cask/cdap/api/artifact/PluginContext.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.artifact;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.templates.AdapterConfigurer;
+import co.cask.cdap.api.templates.ApplicationTemplate;
+import co.cask.cdap.api.templates.plugins.PluginConfig;
+import co.cask.cdap.api.templates.plugins.PluginProperties;
+
+/**
+ * Provides access to plugin context when a program is executing.
+ */
+@Beta
+public interface PluginContext {
+
+  /**
+   * Gets the {@link PluginProperties} associated with the given plugin type and name.
+   *
+   * @param pluginId the unique identifier provide when declaring plugin usage in the program.
+   * @return the {@link PluginProperties}.
+   * @throws IllegalArgumentException if no plugin for the given type and name
+   * @throws UnsupportedOperationException if the program does not support plugin
+   */
+  PluginProperties getPluginProps(String pluginId);
+
+  /**
+   * Loads and returns a plugin class as specified by the given type and name.
+   *
+   * @param pluginId the unique identifier provide when declaring plugin usage in the program.
+   * @param <T> the class type of the plugin
+   * @return the resulting plugin {@link Class}.
+   * @throws IllegalArgumentException if no plugin for the given type and name
+   * @throws UnsupportedOperationException if the program does not support plugin
+   */
+  <T> Class<T> loadClass(String pluginId);
+
+  /**
+   * Creates a new instance of a plugin. The instance returned will have the {@link PluginConfig} setup with
+   * {@link PluginProperties} provided at the time when the
+   * {@link AdapterConfigurer#usePlugin(String, String, String, PluginProperties)} was called during the
+   * {@link ApplicationTemplate#configureAdapter(String, Object, AdapterConfigurer)} time.
+   *
+   * @param pluginId the unique identifier provide when declaring plugin usage in the program.
+   * @param <T> the class type of the plugin
+   * @return A new instance of the plugin being specified by the arguments
+   *
+   * @throws InstantiationException if failed create a new instance
+   * @throws IllegalArgumentException if no plugin for the given type and name
+   * @throws UnsupportedOperationException if the program does not support plugin
+   */
+  <T> T newInstance(String pluginId) throws InstantiationException;
```

Comment:
Usually I don't like the method name repeat the class name. However, given this is a context interface, meaning an implementation might be implementing multiple context interfaces (e.g. DatasetContext, PluginContext... ), having a newInstance method become ambiguous.